### ISSUE TITLE
Dex database migration guidance

### ISSUE THREAD

**Issue by al45tair** [dexidp/dex#1825]
There should be some documentation about the database migrations Dex is likely to attempt.

More specifically, we'd like to be able to do a rolling upgrade of Dex (at least potentially), and/or we'd like to know somehow which versions we can't do that with, so that we minimise the time during which our system can't authenticate users.

If there was a guarantee that Dex would only perform migrations that would be safe to perform while a previous version of Dex was looking at the same database, that would be ideal. Failing that, some kind of guidance about when that won't work (maybe in release notes)?

**Comment by sagikazarmark** [comment]
You are absolutely right! This debt is long overdue. I'd even say operators should have a way to disable automatic migrations, so they have full control over when happens what.

In order to do that, first we need to think about rewriting the SQL layer.

As you can see, migrations at the moment are just SQL queries embedded into Dex: https://github.com/dexidp/dex/blob/a28f5bb218a3186ca55465c4f35572695119c16e/storage/sql/migrate.go#L83
